admin/app/com/lucidchart/piezo/admin/utils/CronHelper.scala

```diff
@@ -0,0 +1,185 @@
+package com.lucidchart.piezo.admin.utils
+
+import java.util.{Date, TimeZone}
+import org.quartz.{CronExpression, CronScheduleBuilder, CronTrigger, TriggerBuilder}
+import play.api.Logging
+import scala.util.control.NonFatal
+
+object CronHelper extends Logging {
+  val DEFAULT_MAX_INTERVAL = 0
+  val NUM_COMPLEX_SAMPLES = 8000
+
+  /**
+   * Determines the max interval for a cron expression. The operation fails silently because it's not crucial.
+   * @param cronExpression
+   */
+  def getMaxInterval(cronExpression: CronExpression): Long = getMaxInterval(cronExpression, None)
+  def getMaxInterval(cronExpression: CronExpression, timeZone: Option[TimeZone]): Long = {
+    try {
+      val subexpressions = cronExpression.getCronExpression.split("\\s")
+      val isComplex = !subexpressions.drop(3).forall(expr => expr == "*" || expr == "?")
+      if (isComplex) getComplexMaxInterval(subexpressions, timeZone) else getSimpleMaxInterval(subexpressions, timeZone)
+    } catch {
+      case NonFatal(e) =>
+        logger.error("Failed to validate cron expression", e)
+        DEFAULT_MAX_INTERVAL
+    }
+  }
+
+  /**
+   * Determines the max interval by deduction. Is only used when seconds, minutes, and hours are specified. If days,
```

Comment:
I think an approach that would be simpler would be to just use your "sampling" method in all cases, and dynamically determine the number how long to sample for.  Basically, find the rightmost/largest unit that isn't `*` or `?`, then iterate over every trigger in the range of that unit. For example, if the largest non-wildcard unit is hours, then get ever trigger for the next day and find the  largest interval.  There would be some situations where this might perform worse than your current deductive logic, like if you had it set to run every second on every other hour of the day, but I think those would be unusual, and we don't really need to worry about them. And it is less likely to run into weird edge cases, and you don't have to worry about more complicated expressions, because the CronExpression class will take care of that for you. It also insures that your logic for parsing the cron expresssion  is more in-sync with how piezo actually handles it.

Here is what I think you would need to run for for each unit:

| Largest specificified unit | how long to get samples for |
:----- | :------
seconds | 1 minute
minutes | 1 hour
hours | 1 day (maybe also run over leap days, if dealing with time zones)
day of month | a little over a month, but make sure to run during a 31 day month rather than current time. May need to handle "W" values specially, since that depends on the month, maybe just add 2 to the result in that case?
month | 1 year
day of week | 7 days normally (note you can't have both day of month and day of week). But if the # mode is used, we'll need to do at least a month, probably more since the exact duration depends on the months involved. Maybe add a buffer of 1 week in that case?
year | this would be pretty unusual, but I guess run from now until 2199? or use some deduction for this one?

Reply:
I think there's a lot of great ideas here. I essentially took a lot of what you said and mixed it with what I had. I now use sampling for each of the subexpressions (seconds, minutes, hours, days). I also group day of month, month, day of week, and year altogether as "days" since all of these can affect the cadence at which days fire.